File: 087. Scramble String/C++/Solution.cpp
class Solution {
public:
    bool isScramble(string s1, string s2) {
        if (s1 == s2) {
            return true;
        }
        if (s1.length() != s2.length()) {
            return false;
        }
        int length = s1.length();
        vector<int> count(26, 0);
        for (int i = 0; i < length; i++) {
            count[s1[i] - 'a']++;
            count[s2[i] - 'a']--;
        }
        for (int item : count) {
            if (item != 0) {
                return false;
            }
        }
        for (int i = 1; i < length; i++) {
            if (isScramble(s1.substr(0, i), s2.substr(0, i)) && isScramble(s1.substr(i), s2.substr(i))) {
                return true;
            }
            if (isScramble(s1.substr(0, i), s2.substr(length - i)) &&
                isScramble(s1.substr(i), s2.substr(0, length - i))) {
                return true;
            }
        }
        return false;
    }
};

File: 087. Scramble String/Python/Solution.py
class Solution(object):
    def isScramble(self, s1, s2):
        """
        :type s1: str
        :type s2: str
        :rtype: bool
        """
        if s1 == s2:
            return True
        if len(s1) != len(s2):
            return False
        length = len(s1)
        count = [0] * 26
        for l1, l2 in zip(s1, s2):
            count[ord(l1) - ord('a')] += 1
            count[ord(l2) - ord('a')] -= 1
        for item in count:
            if item:
                return False
        for i in range(1, length):
            if self.isScramble(s1[:i], s2[:i]) and self.isScramble(s1[i:], s2[i:]):
                return True
            if self.isScramble(s1[:i], s2[length - i:]) and self.isScramble(s1[i:], s2[:length - i]):
                return True
        return False
